### ISSUE TITLE
Remove ServiceTracker from BaseThingHandler

### ISSUE THREAD

**Issue by htreu** [eclipse/smarthome#5182]
The abstract `BaseThingHandler` uses OSGi ServiceTrackers to gain access to the Thing-, Link- and TypeTypeRegistry and to the ConfigDescriptionValidatorService. It also holds a reference to the BundleContext.
We don't want bindings to access these services directly but restrict the access and encapsulate it in the ThingHandlerCallback.

**Comment by lolodomo** [comment]
I search which are the bindings using thingRegistry from `BaseThingHandler` and I found only 2:

- Sonos binding in ESH: it is to get the thing handler of a Sonos group coordinator
- Allplay binding in OH2

I found 3 other bindings in OH2 using directly `ThingRegistry`:

- Powermax binding: it is for the console extension => normal I think
- amazon echo control binding
- neeo IO binding: I don't know why but this binding requires all the ESH registries !

This would be interesting to check if there is not a conception problem with the neeo binding.

**Comment by lolodomo** [comment]
@tmrobert8 : can you comment regarding the neeo binding please ?

**Comment by kaikreuzer** [comment]
The NEEO binding is not a binding in the classical sense - it also includes an IO bundle, i.e. it exposes the entire openHAB to an external system; for this reason it requires access to quite some registries.
It definitely is no code people should look at when they want to learn how to implement a pure binding!

**Comment by tmrobert8** [comment]
@lolodomo - @kaikreuzer pretty succinctly stated the IO portion usage and I don't have much more to add.  Please let me know if you need any more details..

**Comment by lolodomo** [comment]
I found 2 other bindings using the `ThingRegistry` (not through `BaseThingHandler`):

- Zigbee binding: it is for the console extension (like the Powermax binding) => normal I think
- Zwave binding

To comeback to the original topic (removing `thingRegistry` from `BaseThingHandler` or not exposing it), I can find 2 possible workarounds for the Sonos and AllPlay bindings:

1. Declare the link to ThingRegistry in the binding itself
2. Add and expose a method `getAllThings` in the `BaseThingHandler` that will return the list of all things supported by the binding. This solution suppose that `thingRegistry` is not removed from `BaseThingHandler` but only made hidden (private).

So, to be clear, if we want to remove `ThingRegistry` from `BaseThingHandler` quickly, we can first implement the solution 1 to the Sonos and AllPlay bindings and then it can be done.

**Comment by htreu** [comment]
From a framework pov the thing handlers should not depend on the ThingRegistry at all, they must only deal with the thing they were created for.
I now the sonos binding deals with multiple things at once, but this is not the intended way. Instead of the "handler" part of a binding this might in the future be handled by rule actions (see https://github.com/eclipse/smarthome/projects/3#card-8298857).
Since we lack an alternative way of implementing stuff like that we can not further touch this dependency for now.

**Comment by boc-tothefuture** [comment]
If the ThingRegistry is removed from the BaseHandler how are bindings supposed to get access to their Bridge?

**Comment by lolodomo** [comment]
In general, the bridge handler is passed to the thing handler and so you have an access to the bridge from  any thing (of this bridge).

**Comment by boc-tothefuture** [comment]
@lolodomo -
  When you say that 

> In general, the bridge handler is passed to the thing handler

Are you saying that binding implementers are responsible for memoizing the bridge in a ThingHandlerFactory and passing that on in the constructor of any new ThingHandler objects?

or are you saying another mechanism exists?

I am asking, because in the latest code I see the BaseThingHandler uses the registry to get the bridge:
``` 
protected @Nullable Bridge getBridge() {
        ThingUID bridgeUID = thing.getBridgeUID();
        synchronized (this) {
            if (bridgeUID != null && thingRegistry != null) {
                return (Bridge) thingRegistry.get(bridgeUID);
            } else {
                return null;
            }
        }
```

**Comment by lolodomo** [comment]
Ok, you're right, `getBridge()` can be used.

**Comment by cweitkamp** [comment]
> If the ThingRegistry is removed from the BaseHandler how are bindings supposed to get access to their Bridge?

@boc-tothefuture That is the challenge to solve. From my point of view a proper solution removes access to the `ThingRegistry` but not to the bridge itself. After the refactoring the result of the method `getBridge()` should be the same like it is now without having any effect on the bindings.

Maybe your comments leads me to an idea how to implement it. Currently the thing "knows about its bridgeUID". Maybe we should pass a reference to its bridge instead (or the bridge handler, or both).

**Comment by boc-tothefuture** [comment]
@cweitkamp 
It would be much better if we pass a reference to the bridge itself.  It should also be available at construction time, otherwise it leads to the either a lot of null checks, or the need to build a memoizer inside of each binding.

**Comment by htreu** [comment]
> It should also be available at construction time

I don't think this is the case: Things and thing handlers are not created in any particular order. 

> it leads to the either a lot of null checks

thats already the case right now. A bridged thing can not rely on the bridge to be present at any time. Bridge and bridged thing handlers have independent lifecycles.

Since `BaseThingHandler` is a framework class the usage of the service tracker is ugly but necessary (at least I have no better idea right now). The change we would want to introduce here is to switch the `thingRegistry` field from `protected` to `private` access.

**Comment by boc-tothefuture** [comment]
> I don't think this is the case: Things and thing handlers are not created in any particular order.

I thought that changed and it was guaranteed that the bridge was created before any instances of handlers that relied on the bridge?

> thats already the case right now. A bridged thing can not rely on the bridge to be present at any time. Bridge and bridged thing handlers have independent lifecycles.

That is unfortunate - but perhaps I have the wrong mental model of what the bridge is supposed to provide.  In my mental model, it provides a conduit for communication between the things and the device being managed, as well as providing (potentially) some global configuration for things linked to that bridge.

**Comment by htreu** [comment]
> I thought that changed and it was guaranteed that the bridge was created before any instances of handlers that relied on the bridge?

No, things are created in any order. The "child" things get notified about status updates from the bridge in the `bridgeStatusChanged` method.

> but perhaps I have the wrong mental model 

What you describe is correct and is not contradicted by the way handlers are created. Once a bridge handler is created and initialised all child handlers will be informed. Child handlers may set a status of `UNINITIALIZED:BRIDGE_UNINITIALIZED` when no bridge is available.

**Comment by boc-tothefuture** [comment]
>  No, things are created in any order. The "child" things get notified about status updates from the bridge in the bridgeStatusChanged method.

Thanks.  So am I misunderstanding this merged PR?  https://github.com/eclipse/smarthome/pull/2087

Which states:

> A BridgeHandler of a bridge is initialized before ThingHandlers of its child things are initialized.
> A BridgeHandler is disposed after all ThingHandlers of its child things are disposed.

**Comment by htreu** [comment]
Thanks for referencing #2087. I was actually wrong and the PR states that indeed child handlers are created after the bridge handler is present. Sorry for my ignorance and thanks for letting me learn :-)

Coming back to the original topic about removing the ThingRegistry from `BaseThingHandler`: Since our API does not provide means to pass the bridge handler to the HandlerFactory the architecture should stay as it is. The access of ThingRegistry should move to the HandlerCallback and `BaseThingHandler` should delegate the `getBridge` call to the HandlerCallback.

As @cweitkamp stated:

> After the refactoring the result of the method getBridge() should be the same like it is now

Which should be possible through the callback.

Sorry again for causing confusion.
